engine/crates/graphql-schema-diff/src/state.rs

```diff
@@ -266,3 +262,56 @@ fn push_schema_definition_changes(
         }),
     }
 }
+
+fn opt_type_eq(a: Option<&ast::Type<'_>>, b: Option<&ast::Type<'_>>) -> bool {
+    match (a, b) {
+        (Some(a), Some(b)) => type_eq(a, b),
+        (None, None) => true,
+        _ => false,
+    }
+}
+
+fn type_eq(a: &ast::Type<'_>, b: &ast::Type<'_>) -> bool {
+    if a.name() != b.name() {
+        return false;
+    }
+
+    let mut a_wrappers = a.wrappers();
+    let mut b_wrappers = b.wrappers();
+
+    if (&mut a_wrappers).zip(&mut b_wrappers).any(|(a, b)| a != b) {
+        return false;
+    }
+
+    a_wrappers.next().is_none() && b_wrappers.next().is_none()
+}
+
+fn value_eq(a: &ast::Value<'_>, b: &ast::Value<'_>) -> bool {
+    // Same enum variants and equal values
+    match (a, b) {
+        (ast::Value::Int(a), ast::Value::Int(b)) => a == b,
+        (ast::Value::Variable(a), ast::Value::Variable(b)) => a == b,
+        (ast::Value::Float(a), ast::Value::Float(b)) => a == b,
+        (ast::Value::String(a), ast::Value::BlockString(b)) => a == b,
+        (ast::Value::String(a), ast::Value::String(b)) => a == b,
+        (ast::Value::BlockString(a), ast::Value::String(b)) => a == b,
+        (ast::Value::BlockString(a), ast::Value::BlockString(b)) => a == b,
+        (ast::Value::Boolean(a), ast::Value::Boolean(b)) => a == b,
+        (ast::Value::Null, ast::Value::Null) => todo!(),
+        (ast::Value::Enum(a), ast::Value::Enum(b)) => a == b,
+        (ast::Value::List(a), ast::Value::List(b)) => {
+            a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| value_eq(a, b))
+        }
+        (ast::Value::Object(a), ast::Value::Object(b)) => {
+            a.len() == b.len()
+                && a.iter().all(|(name, value)| {
+                    let Some((_, b_value)) = b.iter().find(|(b_name, _)| b_name == name) else {
+                        return false;
+                    };
+
+                    value_eq(value, b_value)
+                })
+        }
+        _ => false,
+    }
+}
```

Comment:
Should these last two be upstreamed in some way?

Reply:
Yes, these seem like they'd be useful.